### PR TITLE
[Papercut][SW-15997] Global Bcc Mail Option

### PR DESCRIPTION
**Pull request by shyim**
## Description

Please describe your pull request:
It adds a global bcc mail fieldin the mail settings

| Questions | Answers |
| --- | --- |
| BC breaks? | no |
| Tests pass? | yes |
| Related tickets? | [Issue](https://issues.shopware.com/#/issues/SW-15997) |
| How to test? | Go to Grundeinstellungen => Shopeinstellungen=> Stammdaten and insert a bcc mail and order a product :) |

### CODE REVIEW

**Comment by janbuecker** [comment]
Is there any reason why you don't use `$this->addBcc()` here? Your implementation isn't really a bcc.

**Comment by shyim** [comment]
@janbuecker i tried addBcc and dont received messages. I can change it back to addBcc

**Comment by bcremer** [comment]
Instead of modifiing `Enlight_Components_Mail_Send` you should add a separate subscriber class that listens to the `Enlight_Components_Mail_Send` event and adds the global BCC options.

File: _sql/migrations/786-add-bcc-mail-field.php [hunk 1]
```diff
@@ -0,0 +1,17 @@
+<?php
+
+class Migrations_Migration786 extends Shopware\Components\Migrations\AbstractMigration
+{
+    public function up($modus)
+    {
+        $sql = <<<'EOD'
+            SET @menu_id = (SELECT id FROM s_core_config_forms WHERE label='Stammdaten');
```
Comment (bcremer) [inline _sql/migrations/786-add-bcc-mail-field.php]: Every Statement has to be a separate call to `addSQL()`. This is due to a long standing bug in PHP's PDO. See: https://bugs.php.net/bug.php?id=61613

**Comment by bcremer** [comment]
@shyim Are you going to change this to an event subscriber as noted in the review? We can also close this PR so the issue is free to pickup for the next person if you prefer that.

File: engine/Library/Enlight/Components/Mail.php [hunk 2]
```diff
@@ -329,6 +335,19 @@ public function send($transport = null)
             )
         );
 
+        $bccs = Shopware()->Config()->get('mailBcc');
```
Comment (hlohaus) [inline engine/Library/Enlight/Components/Mail.php]: pls, don't use "Shopware()" in a Service.
 https://developers.shopware.com/blog/2015/11/11/best-practices-of-shopware-plugin-development/#shopware

**Comment by shyim** [comment]
@hlohaus @bcremer its changed :)

File: engine/Shopware/Components/BccSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            return;
+        }
+
+        $bccs = $this->container->get('config')->get('mailBcc');
+
+        /** @var \Enlight_Components_Mail $mail */
+        $mail = $args->get('mail');
+
+        if (!empty($bccs)) {
+            $mail->addBcc(explode(',', $bccs));
```
Comment (FiveDigital) [inline engine/Shopware/Components/BccSubscriber.php]: Perhaps you should also `trim` the bccs to remove empty spaces.

Comment (shyim) [reply engine/Shopware/Components/BccSubscriber.php]: fixed

File: engine/Shopware/Components/BccSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+            return;
+        }
+
+        $bccs = $this->container->get('config')->get('mailBcc');
+
+        /** @var \Enlight_Components_Mail $mail */
+        $mail = $args->get('mail');
+
+        if (!empty($bccs)) {
+            $bccs = array_map(function ($value) {
```
Comment (janbuecker) [inline engine/Shopware/Components/BccSubscriber.php]: There is no need for an anonymous function here. You can use `array_map('trim', explode(',', $bccs))`

**Pull request merged**